clang/test/CodeGen/ms-cbstring.c

```diff
@@ -0,0 +1,103 @@
+// Comprehensive test for -fms-cbstring flag functionality
+// RUN: %clang_cc1 -emit-llvm -fms-cbstring -fms-extensions %s -o - | FileCheck --check-prefixes=CBSTRING,CHECK %s
+// RUN: %clang_cc1 -emit-llvm -fms-extensions %s -o - | FileCheck --check-prefixes=NORMAL,CHECK %s
+
+// Function declarations to avoid needing system headers
+int printf(const char* format, ...);
+
+// Helper function declaration
+int uses_string(const char*);
+
+// Global scope strings
+const char *global_str = "global_test";
+static const char *static_global_str = "static_global_test";
+
+// String arrays
+const char *array[] = {"array_element1", "array_element2"};
+
+// Function in default text section
+void test_function() {
+    const char *local_str = "local_test";
+    static const char *static_local_str = "static_local_test";
+    
+    // Concatenated strings
+    const char *concat = "concat_" "string_test";
+    
+    // Use strings to prevent optimization
+    printf("%s %s %s\n", local_str, static_local_str, concat);
+}
+
+// Function with custom code segment
+void __declspec(code_seg("other_seg")) in_other_seg(void)
+{
+    uses_string("test_in_other_seg");
+}
+
+int main() {
+    const char *main_str = "main_function_test";
+    static const char *static_main_str = "static_main_test";
+    
+    // Use all strings to ensure they appear in IR
+    printf("%s %s %s %s %s\n", global_str, static_global_str, array[0], array[1], main_str);
+    printf("%s\n", static_main_str);
+    test_function();
+    in_other_seg();
+    
+    return 0;
+}
+
+// Check string constants in order they appear in IR:
+// Note: With -fms-cbstring, function strings get section attributes but remain comdats for proper deduplication
+// Note: Global strings keep default section behavior (like MSVC's CONST segment)
+
+// Global strings - should keep default section behavior (not go to .text$s)
+// NORMAL: constant {{.*}} c"global_test\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"global_test\00"{{.*}}comdat{{.*}}align
+// CBSTRING-NOT: constant {{.*}} c"global_test\00"{{.*}}section
```

Comment:
```suggestion
// CHECK: constant {{.*}} c"global_test\00"
// CHECK-NOT: section
// CHECK-SAME: comdat
```

Otherwise `-NOT` checks that the lines doesn't appear AFTER the last `CHECK`, so this isn't a useful check. Also, the `{{.*}}` will eat any section declaration, so this will still match if there is an explicit section.